### ISSUE TITLE
Generic functions

### ISSUE THREAD

**Issue by jarble** [pfusik/cito#4]
Is it possible to define generic classes and functions in Ć, like in Java, C#, and C++?

If Ć doesn't have generics yet, then the syntax could be based on C#, like this:

```
public class GenericClass<T>
{
    //source code goes here
}
```

**Comment by pfusik** [comment]
Generics aren't available yet, aside from the built-in collection types `List<T>`, `Dictionary<TKey, TValue>` and `SortedDictionary<TKey, TValue>`.
Generics weren't available in the first releases of Java and C# either. They ended up being implemented differently.
Could you please describe your usecase for generics?

**Comment by jarble** [comment]
I'm planning to write a compiler that translates TypeScript to Ć, so the translation would be easier if generics were available in this language.

Since Ć targets several languages that already have generics, it might be a relatively easy feature to implement.

**Comment by pfusik** [comment]
I'd like to have generics in Ć, of course!

However, your example translates easily to C++, but not C# or Java. It uses duck typing. C# or Java won't allow arithmetic on generic types.

BTW. do you think adding a TypeScript backend to `cito` would be useful? What will be the advantages versus JavaScript?

**Comment by jedwards1211** [comment]
@pfusik outputting TypeScript declarations (`.d.ts` files) would be great for writing an in API Ć because the TypeScript defs would provide better Intellisense in IDEs than JavaScript.  Anytime I'm using an API it's nice to have all the Intellisense I can get.

And for someone who wants to use the API in a TypeScript project, they would get better compile-time checks.

Outputting full TypeScript source (not just type declarations) wouldn't be as worthwhile though since then you would have to compile those to JS separately.

**Comment by jarble** [comment]
The implementation of generics is much more difficult than I expected: if the goal of this compiler is to target as many languages as possible, then this feature should probably be left out for now.
A TypeScript backend should be relatively easy to implement, since it has many features in common with C# and Java.

**Comment by jedwards1211** [comment]
@jarble yeah I assume there are some tricky differences between C++ templates and generics.

**Comment by jedwards1211** [comment]
@jarble I'm working on TypeScript declarations output: #13

**Comment by pfusik** [comment]
Back to the topic. I'd like to see real code examples where generics are useful. From my experience, it's mostly collections. And yes, some collections are not implemented yet, most notably `Set`.

**Comment by pfusik** [comment]
I've added `Stack<T>` and `HashSet<T>`.

**Comment by pfusik** [comment]
[Go introduces generics](https://go.dev/blog/why-generics) after 12 years.

**Comment by pfusik** [comment]
Added `OrderedDictionary<TKey, TValue>` and `Queue<T>`.

**Comment by charliefoxtwo** [comment]
> Back to the topic. I'd like to see real code examples where generics are useful. From my experience, it's mostly collections. And yes, some collections are not implemented yet, most notably `Set`.

I decided to try out writing an api client in fusion so that it could be made available for many languages. In the original typescript client, we have a response type like this (simplified):
```ts
interface Result<T> {
  success: boolean;
  data?: T;
  errorMessage?: string;
  statusCode: number;
}
```
This allows us to have one common type that we can return for calls to any endpoint, with a strongly-typed data object. Without generics, wouldn't I need a new return type for each method? Or I could have all of my data objects extend a base class but that doesn't feel like the right solution here.

(as an aside, lack of generics also creates issues with json parsing where I now have to have a type assertion in every single method instead of having something like Parse<T> which can handle that for me).

This sort of Response/Result object is a very common pattern for generics, similar to Java's Optional class. It would be great to be able to support something like this (though I understand it wouldn't be easy).

**Comment by pfusik** [comment]
I need more context: what are the "data objects" ? What are the methods? Please show more code.

> Or I could have all of my data objects extend a base class but that doesn't feel like the right solution here.

Why not? If these objects have common use (they are part of `Result`), the OOP way is to make them derive from a common class.

> I now have to have a type assertion

Could you give examples? Type assertions seem to be TypeScript-specific.

Regarding JSON: there's an experimental parser for D, JavaScript/TypeScript, Python, Swift and a compatible pure-Fusion implementation for C, C++ and Java: #140. This all works without generics. I don't see how generics would help here.

**Comment by charliefoxtwo** [comment]
> Why not? If these objects have common use (they are part of Result), the OOP way is to make them derive from a common class.

The contents `data` should not necessarily include the contents of `Result`.

Consider this simple example in C#:
```cs
public record Foo(int A, string B);
public record Bar(bool C, double D, Foo E);
public record Result<T>(bool Success, T Data);

interface IApiClient
{
    Result<Foo> GetFoo();
    Result<Bar> GetBar();
}
```

`Foo` cannot (and should not) inherit `Result`, because it is also used in `Bar` where it should not have the `Success` property.

> Could you give examples? Type assertions seem to be TypeScript-specific.

The JSON parser example shared is cool (and seems like it would be a great addition!), but would it not be convenient to be able to do something like this:
```cs
Foo foo = parser.Parse<Foo>(fooJson);
Bar bar = parser.Parse<Bar>(barJson);
```
That way the caller of the library doesn't have to handle the `JsonObject` themselves? Or perhaps I missed something in the example that makes that possible? The only way I've been able to come up with is to have the caller pass in their own parser which will return an object (I couldn't find an `object` type in Fusion, and since generics are not supported I had to settle for an empty base class for testing purposes). This means every request method has to have this type casting to return the desired type to the caller, which feels repetitive right?
```
public abstract class JsonSerializer
{
    public abstract string Serialize(RequestBody obj);
    public abstract ResponseBody Deserialize(string str);
}

public abstract class ResponseBody
{
}

public ResponseObject MakeRequest(RequestObject x) throws Exception
{
    string body = _jsonSerializer.Serialize(x);
    string response = _httpClient.MakeRequest(body);
    if (_jsonSerializer.Deserialize(response) is ResponseObject ro)
    {
        return ro;
    }
    throw Exception();
}
```

Maybe this is all just because I'm so used to generics, but I'm really not seeing another way to do this without them?
